README
MP3::M3U::Parser
================

Parses M3U mp3 playlists and if wanted, exports the parsed data to 
formats like xml and html.

Read the module's POD for documentation.

See the "t" and "eg" directories for examples.

INSTALLATION

Just use the cpan shell to install this module. From a command prompt, type:

    cpan

or

    perl -MCPAN -e shell

to enter the cpan shell. And then enter:

   install MP3::M3U::Parser

DEPENDENCIES

This module requires these other modules:

   Text::Template

COPYRIGHT AND LICENCE

See the LICENSE file bundled with this distribution.


Changes
Revision history for Perl extension MP3::M3U::Parser.

2.31 Sun Jul  8 01:50:04 2012
    => Bump version.

2.30 Thu Dec 31 04:39:58 2009
    => Perl::Critic refactoring.
    => HTML::Entities is now an optional dependency.
    => New tests.

2.24 Thu Sep  3 02:02:02 2009
    => Pod & distro fixes. Also correct a typo after __END__.

2.23 Thu Apr 23 14:50:16 2009
    => Update builder.

2.22 Sat Apr 18 06:23:24 2009
    => Update distro tools.

2.21 Sun Mar 29 08:20:01 2009
    => Fixed an issue related to network paths.
       Reported by Paul Miller <jettero@cpan.org>.
    => Converted the distro files to LF from CRLF.
    => Updated test suite.
    => Test::More 0.40 is now required.

2.20 Sun Jun 18 03:58:36 2006
    => SIGNATURE sent to oblivion :p
    => Added Build.PL
    => Updated Makefile.PL
    => Updated tests
    => Updated copyright & license
    => Private methods (including template() & search())
       renamed to include an initial underscore:
          search() -> _search()
          template -> _template()
          etc.
       Beware() if you->are('sub-classing');

2.1 Tue Sep 14 20:47:09 2004
    => export() now has the ability to dump the generated data
       to a scalar instead of exporting to a file. Useful for
       web programming and dynamic generation. To enable this,
       -toscalar option must be passed with a scalar ref.
    => Module now require()s IO::File, File::Spec and Cwd if needed.
       Because, we won't need this modules, if the user works with
       scalars instead of disk files. This will save some memory 
       and incerase speed I believe. If you are parsing scalar
       data and do not use export(), then no external modules 
       will be loaded.
    => Updated HTML and XML templates for -toscalar compatibility.
    => Updated Makefile.PL
    => Updated pod.
    => Added new tests to demonstrate -toscalar option.
    => Added a CGI example code to distro.
    => Some minor fixes.

2.01 Mon Jul 26 19:54:14 2004
    => Minor fix in Pod.
    => SONG element was including ".mp3" if it is 
       extracted from PATH. Fixed.
    => ARTIST element was including path info. Fixed.
       But, if the song name or artist includes slash or backslash,
       we have a problem

2.0 Fri Jul 23 23:19:48 2004
    => BEWARE!!! THIS NEW VERSION BREAKS THE OLD CODE!
    => Now the module looks *like* a module. It was
       like a program code before. With this version
       it does only the parsing job. 
    => Removed directory reading codes. 
    => Removed some parameters passed to new()
    => Removed ignore chars thingies.
    => Removed Dump() method.
    => parse() now accepts file_paths/filehandles/scalars as 
       parameters. You can mix and use these types together; the 
       parser is now smart enough to do the right job with the 
       related type.
    => parse() now returns the object itself. 
       Use the brand new result() method to get the parsed tree.
       This new behaviour is a result of adding the chaining
       ability of public methods.
    => Object table structure and m3u key orders are changed.
       Object table is now more complex and verbose.
    => Returned result set is now an array not hash.
    => Module now checks if the file is a valid m3u file, 
       before parsing it. Dies if it is invalid.
    => The new method reset() resets the parser object. Might
       be usefull in loops. For example: you can export each
       m3u to it's separate file with reset() without any extra work.
    => The module no longer requires Data::Dumper, but it now requires
       Cwd and Text::Template.
    => Exported HTML file is now fully templated. You can subclass
       the module and override template() method to change the HTML 
       template. Currently, I'm not planning to add a template for 
       XML type.
    => Added new namespace MP3::M3U::Parser::Dummy for Text::Template.
    => Added tcompile() method for compiling templates.
    => drive key returned from info() is now an array.
    => Added locate_file() method to get the full path of a file,
       if the file is in the current working directory.
    => Parser now tries to extract the artist and song infos into 
       a separate data field.
    => New option '-overwrite' added to new() and export(). It 
       controls the behaviour of module in export() globally or 
       locally depending on where you set it.
    => export() checks if the export file already exists. Dies
       if it exists and you didn't select to overwrite it.
    => Pod updated.
    => Subclassing examples added to Pod.
    => You can now set the encoding of exported files 
       from new() globally via -encoding option.
    => You can now set the format of exported files from 
       new() globally via -expformat option.
    => You can now set the -drives option for export() from 
       new() globally via -expdrives option.
    => But parameters passed to export() has the priority.
    => XML element names changed.

1.1 Mon Aug 25 21:42:31 2003
    => Better XML escaping.
    => Added '-ignore_chars' parameter.
    => Fixed a bug in seconds()
    => Moved module to ./lib/MP3/M3U in the 
       distribution package.

1.04 Sun Jun 22 15:54:43 2003
    => Removed sort() from export() method (the module was using 
       it on the $self->{M3U}{$cd} object table). 
       Thanks to Patrick Bourdon.

1.03 Tue Jun 10 22:38:53 2003
    => HTML code is valid HTML 4.01 & valid CSS. But you can get 
       "non SGML character number" errors if your list(s) have 
       some of the undefined chars. I may add a filter to 
       escape these in the future.
    => Some minor fixes.

1.02 Mon Mar 10 22:33:04 2003
    => Fixed the comma bug. Reported by Brandon Lederer.
    => Fixed a bug in the formatting seconds part.
    => Empty result keys (if a search didn't find any matches 
       in a list for example) are ignored when exporting to xml.

lib/MP3/M3U/Parser/Export.pm
package MP3::M3U::Parser::Export;
use strict;
use warnings;
use vars qw( $VERSION );
use Carp qw( croak );
use MP3::M3U::Parser::Constants;
use MP3::M3U::Parser::Dummy;

$VERSION = '2.31';

my %DEFAULT = (
    format    => 'html',
    filename  => 'mp3_m3u%s.%s',
    encoding  => 'ISO-8859-1',
    drives    => 'on',
    overwrite => 0,
    toscalar  => 0,
);

sub export {
    my($self, @args) = @_;
    my %opt       = @args % 2 ? () : @args;
    my $format    = $opt{'-format'}    || $self->{expformat}   || $DEFAULT{format   };
    my $encoding  = $opt{'-encoding'}  || $self->{encoding}    || $DEFAULT{encoding };
    my $drives    = $opt{'-drives'}    || $self->{expdrives}   || $DEFAULT{drives   };
    my $overwrite = $opt{'-overwrite'} || $self->{overwrite}   || $DEFAULT{overwrite};
    my $to_scalar = $opt{'-toscalar'}  || $self->{exptoscalar} || $DEFAULT{toscalar };
    my $file      = $opt{'-file'}      || $self->_default_filename( $format );

    $file = $self->_locate_file($file) if ! $to_scalar;
    my $OUTPUT = $format eq 'xml'
               ? $self->_export_to_xml(  $encoding )
               : $self->_export_to_html( $encoding, $drives, $to_scalar, $file)
               ;

    if ( $to_scalar ) {
        ${$to_scalar} = $OUTPUT;
    }
    else {
        my $fh = $self->_check_export_params( $file, $to_scalar, $overwrite );
        print {$fh} $OUTPUT or croak "Can't print to FH: $!";
        $fh->close;
    }

   $self->{EXPORTF}++;
   return $self if defined wantarray;
   return;
}

sub _default_filename {
    my($self, $format) = @_;
    croak 'Export format is missing' if ! $format;
    return sprintf $DEFAULT{filename}, $self->{EXPORTF}, $format;
}

sub _check_export_params {
    my($self, $file, $to_scalar, $overwrite) = @_;
    my $fh;
    if ( $to_scalar && ( ! ref $to_scalar || ref $to_scalar ne 'SCALAR' ) ) {
        croak '-toscalar must be a SCALAR reference';
    }
    if ( ! $to_scalar ) {
        if ( -e $file && ! $overwrite ) {
           croak "The export file '$file' exists & overwrite option is not set";
        }
        require IO::File;
        $fh = IO::File->new;
        $fh->open( $file, '>' )
            or croak "I can't open export file '$file' for writing: $!";
    }
    return $fh;
}

sub _export_to_html {
    my($self, $encoding, $drives, $to_scalar, $file) = @_;
    my $OUTPUT = EMPTY_STRING;
    # I don't think that weird numbers in the html mean anything 
    # to anyone. So, if you didn't want to format seconds in your 
    # code, I'm overriding it here (only for export(); Outside 
    # export(), you'll get the old value):
    my $old_seconds = $self->{seconds};
    $self->{seconds} = 'format';
    my %t;
    @t{ qw( up cd data down ) } = split m{\Q<!-- MP3DATASPLIT -->\E}xms,
                                        $self->_template;
    foreach (keys %t) {
        $t{$_} = $self->_trim( $t{$_} );
    }
    my $tmptime = $self->{TOTAL_TIME} ? $self->_seconds($self->{TOTAL_TIME})
                :                       undef;
    my @tmptime;

    if ($tmptime) {
        @tmptime = split m{:}xms,$tmptime;
        unshift @tmptime, 'Z' if $#tmptime <= 1;
    }

    my $average = $self->{AVERAGE_TIME}
                ? $self->_seconds( $self->{AVERAGE_TIME} )
                : '<i>Unknown</i>'
                ;

    my $HTML = {
        ENCODING    => $encoding,
        SONGS       => $self->{TOTAL_SONGS},
        TOTAL       => $self->{TOTAL_FILES},
        AVERTIME    => $average,
        FILE        => $to_scalar ? EMPTY_STRING : $self->_locate_file($file),
        TOTAL_FILES => $self->{TOTAL_FILES},
        TOTAL_TIME  => @tmptime ? [ @tmptime ]   : EMPTY_STRING,
    };

    $OUTPUT .= $self->_tcompile(template => $t{up}, params=> {HTML => $HTML});
    my($song,$cdrom, $dlen);
    foreach my $cd (@{ $self->{'_M3U_'} }) {
        next if($#{$cd->{data}} < 0);
        $cdrom .= "$cd->{drive}\\" if $drives ne 'off';
        $cdrom .= $cd->{list};
        $OUTPUT .= sprintf $t{cd}."\n", $cdrom;
        foreach my $m3u (@{ $cd->{data} }) {
            $song = $m3u->[ID3];
            if ( ! $song ) {
                my @test_path = split /\\/xms, $m3u->[PATH];
                my $tp        = pop @test_path || $m3u->[PATH];
                my @test_file = split /\./xms, $song;
                $song         = $test_file[0] || $tp;
            }
            $dlen = $m3u->[LEN] ? $self->_seconds($m3u->[LEN]) : '&nbsp;';
            $song = $song       ? $self->_escape($song)        : '&nbsp;';
            $OUTPUT .= sprintf "%s\n", $self->_tcompile(
                                            template => $t{data},
                                            params   => {
                                                data => {
                                                    len  => $dlen,
                                                    song => $song,
                                                }
                                            }
                                        );
        }
        $cdrom = EMPTY_STRING;
    }
    $OUTPUT .= $t{down};
    $self->{seconds} = $old_seconds; # restore
    return $OUTPUT;
}

sub _export_to_xml {
    my($self, $encoding) = @_;
    my $OUTPUT = EMPTY_STRING;
    $self->{TOTAL_TIME} = $self->_seconds($self->{TOTAL_TIME})
                                if $self->{TOTAL_TIME} > 0;
    $OUTPUT .= sprintf qq~<?xml version="1.0" encoding="%s" ?>\n~, $encoding;
    $OUTPUT .= sprintf qq~<m3u lists="%s" songs="%s" time="%s" average="%s">\n~,
                       $self->{TOTAL_FILES},
                       $self->{TOTAL_SONGS},
                       $self->{TOTAL_TIME},
                       $self->{AVERAGE_TIME};
    my $sc = 0;
    foreach my $cd (@{ $self->{'_M3U_'} }) {
        $sc = $#{$cd->{data}}+1;
        next if ! $sc;
        $OUTPUT .= sprintf qq~<list name="%s" drive="%s" songs="%s">\n~,
                            $cd->{list},
                            $cd->{drive},
                            $sc;
        foreach my $m3u (@{ $cd->{data} }) {
            $OUTPUT .= sprintf qq~<song id3="%s" time="%s">%s</song>\n~,
                                $self->_escape( $m3u->[ID3] ) || EMPTY_STRING,
                                $m3u->[LEN]                   || EMPTY_STRING,
                                $self->_escape( $m3u->[PATH] );
        }
        $OUTPUT .= "</list>\n";
        $sc = 0;
    }
    $OUTPUT .= "</m3u>\n";
    return $OUTPUT;
}

# compile template
sub _tcompile {
    my($self, @args) = @_;
    my $class = ref $self;
    croak 'Invalid number of parameters' if @args % 2;
    require Text::Template;
    my %opt = @args;
    my $t   = Text::Template->new(
                TYPE       => 'STRING',
                SOURCE     => $opt{template},
                DELIMITERS => ['<%', '%>'],
            ) or croak "Couldn't construct the template: $Text::Template::ERROR";

    my @globals;
    foreach my $p ( keys %{ $opt{params} } ) {
        my $ref    = ref $opt{params}->{$p};
        my $prefix = $ref eq 'HASH'  ? q{%}
                   : $ref eq 'ARRAY' ? q{@}
                   :                   q{$}
                   ;
        push @globals, $prefix . $p;
    }

    my $text = $t->fill_in(PACKAGE => $class . '::Dummy',
                PREPEND => sprintf('use strict;use vars qw[%s];',
                                    join q{ }, @globals ),
                HASH    => $opt{params},
              ) or croak "Couldn't fill in template: $Text::Template::ERROR";
    return $text;
}

# HTML template code
sub _template {
   return <<'MP3M3UPARSERTEMPLATE';
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
   "http://www.w3.org/TR/html4/loose.dtd">
<html>
 <head>
   <title>MP3::M3U::Parser Generated PlayList</title>
   <meta name="Generator" content="MP3::M3U::Parser">
   <meta http-equiv="content-type"
         content="text/html; charset=<%$HTML{ENCODING}%>">

   <style type="text/css">
   <!--
      body   { background   : #000040;
               font-family  : font1, Arial, serif;
               color        : #FFFFFF;
               font-size    : 10pt;    }
      td     { background   : none;
               font-family  : Arial, serif;
               color        : #FFFFFF;
               font-size    : 13px;    }
      hr     { background   : none;
               color        : #FFBF00; }
      .para1 { margin-top   : -42px;
               margin-left  : 350px;
               margin-right : 10px;
               font-family  : font2, Arial, serif;
               font-size    : 30px; 
               line-height  : 35px;
               background   : none;
               color        : #E1E1E1;
               text-align   : left;    }
      .para2 { margin-top   : 15px;
               margin-left  : 15px;
               margin-right : 50px;
               font-family  : font1, Arial Black, serif;
               font-size    : 50px;
               line-height  : 40px;
               background   : none;
               color        : #004080;
               text-align   : left;    }
      .t     { font-family  : Arial, serif;
               background   : none;
               color        : #FFBF00;
               font-size    : 13px;    }
      .ts    { font-family  : Arial, serif;
               color        : #FFBF00;
               background   : none;
               font-size    : 10px;    }
      .s     { font-family  : Arial, serif;
               background   : none;
               color        : #FFFFFF;
               font-size    : 13px;    }
      .info  { font-family  : Arial, serif;
               background   : none;
               color        : #409FFF;
               font-size    : 10px;    }
      .infob { font-family  : Arial, serif;
               background   : none;
               color        : #FFBF00;
               font-size    : 15px;    }
    -->
   </style>

 </head>

<body>

 <div align="center">
  <div class="para2" align="center"><p>MP3::M3U::Parser</p></div>
  <div class="para1" align="center"><p>playlist</p></div>
 </div>

<hr align="left" width="90%" noshade="noshade" size="1">
 <div align="left">

  <table border="0" cellspacing="0" cellpadding="0" width="98%">
   <tr><td>
    <span class="ts"><%$HTML{SONGS}%></span> <span class="info"> tracks and 
    <span class="ts"><%$HTML{TOTAL}%></span> Lists in playlist, 
      average track length: </span> 
      <span class="ts"><%$HTML{AVERTIME}%></span><span class="info">.</span>
     <br>
    <span class="info">Playlist length: </span><%
   my $time;
   if ($HTML{TOTAL_TIME}) {
      my @time = @{$HTML{TOTAL_TIME}};
      $time = qq~<span class="ts"  > $time[0] </span>
                 <span class="info"> hours    </span>~ if $time[0] ne 'Z';
      $time .= qq~
            <span class="ts"  > $time[1] </span>
            <span class="info"> minutes  </span>
            <span class="ts"  > $time[2] </span>
            <span class="info"> seconds. </span>~;
   } else {
      $time = qq~<span class="ts"><i>Unknown</i></span><span class="info">.</span>~;
   }
   $time;

     %><br>
    <% 
      qq~<span class="info">Right-click <a href="file://$HTML{FILE}">here</a>
      to save this HTML file.</span>~ if $HTML{FILE}
    %>
    </td>
   </tr>
 </table>

</div>
<blockquote>
<p><span class="infob"><big><% 
$HTML{TOTAL_FILES} > 1 ? "Playlists and Files" : "Playlist files"; 
%>:</big></span></p>

<table border="0" cellspacing="1" cellpadding="2">

<!-- MP3DATASPLIT -->
<tr><td colspan="2"><b>%s</b></td></tr>
<!-- MP3DATASPLIT -->
<tr><td><span class="t"><%$data{len}%></span></td><td><%$data{song}%></td></tr>
<!-- MP3DATASPLIT -->

  </table>
</blockquote>
<hr align="left" width="90%" noshade size="1">
<span class="s">This HTML File is based on 
<a href="http://www.winamp.com">WinAmp</a>`s HTML List.</span>
</body>
</html>
MP3M3UPARSERTEMPLATE
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

MP3::M3U::Parser::Export - Exports playlist to HTML/XML

=head1 SYNOPSIS

Private module.

=head1 DESCRIPTION

This document describes version C<2.31> of C<MP3::M3U::Parser::Export>
released on C<9 September 2012>.

-

=head1 METHODS

=head2 export

See C<export> in L<MP3::M3U::Parser>.

=head1 SEE ALSO

L<MP3::M3U::Parser>.

=head1 AUTHOR

Burak Gursoy <burak@cpan.org>.

=head1 COPYRIGHT

Copyright 2003 - 2012 Burak Gursoy. All rights reserved.

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.3 or,
at your option, any later version of Perl 5 you may have available.
=cut
